New block: Query filter

ryelle opened the pull request:
See #419 — this adds a new block called "Query Filter", which renders a dropdown form using values passed in via filter. When "Apply" is clicked, the form is submitted, and the resulting URL is expected to handle the applied filters. In many cases, wp_query can handle combined filters without any customization (ex, taxonomies). For non-core filters, hooking into `pre_get_posts` (+ maybe `query_vars`?) in the child theme should be enough.

This block is using the [Interactivity API](https://github.com/WordPress/gutenberg/tree/trunk/packages/interactivity), which uses `data-wp-` attributes to attach events and conditional attributes. This API is still experimental, but I took a lot of inspiration from [the navigation block](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/navigation).

By itself, this block doesn't render anything. It expects a `key` attribute, which is used to create a specific filter for the individual sites to populate a settings object. For example, the following would be used to create an "example" filter.

```html
<!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap","justifyContent":"right"}} -->
<div class="wp-block-group">
    <!-- wp:wporg/query-filter {"key":"example"} /-->
</div>
<!-- /wp:group -->
```

This creates the WP filter hook `wporg_query_filter_options_example`, which expects an object used to render the options. There's [more documentation inline in render.php](https://github.com/WordPress/wporg-mu-plugins/blob/0f2c391b8d8ff47c04b5233dee8e76cd2d7bd810/mu-plugins/blocks/query-filter/render.php#L10-L24).

To create a simple dropdown with 3 options:

```php
add_filter(
	'wporg_query_filter_options_example',
	function() {
		return array(
			'label' => 'Example <span>1</span>',
			'key' => 'q',
			'action' => home_url(),
			'options' => array(
				'opt-1' => 'Option 1',
				'opt-2' => 'Option 2',
				'opt-3' => 'Option 3',
			),
			'selected' => array( 'opt-2' ),
		);
	}
);
```

The output would look like this:

<img width="390" alt="Screenshot 2023-09-05 at 6 51 56 PM" src="https://github.com/WordPress/wporg-mu-plugins/assets/541093/85113b1d-cb64-433e-996e-3ab7c7d0792b">

When "Apply" is clicked, the page loads `/?q[]=opt-2`. It's up to each site to load the correct archive.

❓  I think this approach will work well for the child theme setup, but I definitely want feedback on it before committing us all to this for all the directories 🙂 

🖌️  I've also used block-specific custom properties so that other themes can easily change colors with `theme.json` changes (here's an example in showcase, though i removed it once everything was set up since the defaults work). This is in case the theme directory needs a dark scheme.

**Practical example: Showcase**

I've got this working with real filters in https://github.com/WordPress/wporg-showcase-2022/pull/147 — here are some screen recordings of everything in place:

A basic filtering, starting on the homepage, shows all 3 filters & options. When a filter is applied, the results load using the archive template.

https://github.com/WordPress/wporg-mu-plugins/assets/541093/46b6268d-23b8-429a-976f-99af1fe7637e

More filtering interactions — clearing filters using "Clear" and by clicking the option; applying multiple filters, clicking out before applying.

https://github.com/WordPress/wporg-mu-plugins/assets/541093/a231e149-e4b9-40df-b6b7-1d1578bec859

Search and filters can work together — once on a search results page, filtering will filter the search results. (the opposite is not true, if you search on a filtered page it will remove the filters, we talked through that and it was expected behavior)

https://github.com/WordPress/wporg-mu-plugins/assets/541093/277c686f-efb8-4d03-856a-48fdebfb30a9

When clicking on a tag archive page, the filter for that tag is correctly selected.

https://github.com/WordPress/wporg-mu-plugins/assets/541093/54106695-e564-4c32-82d1-6ba447993cba

And on mobile, the dropdown switches to a modal

![Screen Shot 2023-09-05 at 18 37 24](https://github.com/WordPress/wporg-mu-plugins/assets/541093/23b51845-92eb-4d8a-b75c-29e11c6638d9)

ryelle committed:
Add query filter block

ryelle committed:
Refactor filters block

ryelle committed:
Update styles

ryelle committed:
Add action to inject extra field values

ryelle committed:
Rename filter & improve documentation

ryelle committed:
Update filter destination to use value from settings

ryelle committed:
Rename variable

ryelle committed:
Remove color configuration in favor of theme.json settings

ryelle committed:
Create and use more custom theme.json values

ryelle committed:
Set some defaults for custom properties

ryelle committed:
Clear the form using the "clear" button

ryelle committed:
Update the bubble as items are selected, but clear changes when closed without applying

ryelle committed:
Use the `closeDropdown` helper to clear the form

ryelle committed:
Add mobile styles

ryelle committed:
Add key as an expected attribute

StevenDufresne commented on mu-plugins/blocks/query-filter/render.php:
```diff
@@ -0,0 +1,127 @@
+				/>
+				<input
+					type="submit"
+					value="<?php esc_html_e( 'Apply', 'wporg' ); ?>"
```
I think it was agreed that this would be updated with the filter count:

[Context Link](https://github.com/WordPress/wporg-mu-plugins/issues/419#issuecomment-1679390461)
[Figma Link](https://www.figma.com/file/HOJTpCFfa3tR0EccUlu0CM/WordPress.org-Design-Library?type=design&node-id=2847-4389&mode=design&t=Y2B54ZHq9soTX1nE-0)

| Image |
| --- |
| <img width="200" alt="Screenshot 2023-09-06 at 10 22 42 AM" src="https://github.com/WordPress/wporg-mu-plugins/assets/1657336/e04fdba4-676d-4c33-99e9-874a549ae778"> |

ryelle replied:
I've been only referencing the "filters 6.1" page in the Filter component file, thinking that was the final version 🙃 

The link doesn't work for me (opens the file but wrong page)… but I found it by searching through the other pages, so thanks for pointing it out.

@WordPress/meta-design — Can you confirm how this should behave? It sounds like the top number (on black, red highlight) would be the number of filters currently applied on the page, while the Apply button (on blue, yellow highlight) would be the number of filters currently selected— so it would stay "Popular tags 1" even if you select two more, but the button would be "Apply (3)"?

<img width="300" alt="Screenshot 2023-09-06 at 10 50 06 AM" src="https://github.com/WordPress/wporg-mu-plugins/assets/541093/f9e03b81-4938-446d-932f-a5e2cbb93d64">

jasmussen replied:
> It sounds like the top number (on black, red highlight) would be the number of filters currently applied on the page, while the Apply button (on blue, yellow highlight) would be the number of filters currently selected— so it would stay "Popular tags 1" even if you select two more, but the button would be "Apply (3)"?

Correct, but honestly I think we can remove the number on the Apply button for now, I'm not sure it adds clarity. If Francisco when he comes back from AFK has to correct me, I will buy him an apology milkshake at that time. How does this sound?

ryelle replied:
I think it was an accessibility request, but maybe it would be easier to understand the UX once this is merged and live on the staging site. I'll leave this as-is (no number on button) for this initial pass, and ask for feedback once there's a live place to try it out.

fcoveram replied:
@ryelle is correct. It was an accessibility request shared [here](https://github.com/WordPress/wporg-mu-plugins/issues/419#issuecomment-1679390461).

StevenDufresne commented on mu-plugins/blocks/query-filter/postcss/style.pcss:
```diff
@@ -0,0 +1,314 @@
+		font-weight: 400;
+
+		&:hover,
+		&:active {
```
Should this have a `:focus` state?

ryelle replied:
It inherits the focus from the focus just above here, it doesn't need a custom focus style— but if design wants to create one I can change it.

<img width="394" alt="Screenshot 2023-09-06 at 11 01 10 AM" src="https://github.com/WordPress/wporg-mu-plugins/assets/541093/65cab3c8-0222-45e6-adfc-ccdfea7ea7df">

StevenDufresne commented:
I tested it locally, and while I don't yet have the expertise to review the interactivity implementation, it does work well and updates the query string appropriately. 👍

adamwoodnz commented on mu-plugins/blocks/query-filter/render.php:
```diff
@@ -0,0 +1,127 @@
+$encoded_state = wp_json_encode( [ 'wporg' => [ 'queryFilter' => $init_state ] ] );
+
+// Set up a unique ID for this filter.
+$html_id = wp_unique_id( "filter-{$settings['key']}-" );
```
TIL

adamwoodnz commented on mu-plugins/blocks/query-filter/src/view.js:
```diff
@@ -0,0 +1,136 @@
+	}
+	toggle.querySelector( 'span' ).innerText = count;
+	if ( count ) {
+		toggle.classList.add( 'has-filter-applied' );
```
Doesn't look like `.has-filter-applied` is used for anything, maybe not required?

ryelle replied:
👍🏻

adamwoodnz commented on mu-plugins/blocks/query-filter/src/view.js:
```diff
@@ -0,0 +1,136 @@
+					context.wporg.queryFilter.hasHover = true;
+				},
+				handleMouseLeave: ( { context } ) => {
+					context.wporg.queryFilter.hasHover = null;
```
Is there a reason not to always keep this as a boolean?

ryelle replied:
There wasn't, but I've removed this mouse behavior to fix some other glitches I had noticed https://github.com/WordPress/wporg-mu-plugins/pull/441/commits/4f610db003632cb489041ae76e07a4c585c238fa.

ryelle commented on mu-plugins/blocks/query-filter/postcss/style.pcss:
```diff
@@ -0,0 +1,314 @@
+
+	&.wporg-query-filter__modal-action-clear {
+		background-color: transparent;
+		color: #a13f2a; /* pomegrade-1 fails contrast against white. */
```
@WordPress/meta-design I meant to flag this in the description— the mockup uses "Pomegrade" on white but that [fails to meet contrast requirements](https://webaim.org/resources/contrastchecker/?fcolor=E26F56&bcolor=FFFFFF). There's a note in the design library about darker versions of colors, but I didn't see any, so I've made this one up. If you'd rather something different, it just needs to have [a 4.5+ value against light-grey-2](https://webaim.org/resources/contrastchecker/?fcolor=A13F2A&bcolor=F6F6F6), since I'm using that as the hover color.

<img width="358" alt="Screenshot 2023-09-06 at 11 02 36 AM" src="https://github.com/WordPress/wporg-mu-plugins/assets/541093/389c6048-8246-4144-8538-dcfd74c04af1">

Hover:
<img width="376" alt="Screenshot 2023-09-06 at 11 02 50 AM" src="https://github.com/WordPress/wporg-mu-plugins/assets/541093/cf67c069-0d4d-4bf1-8517-48c26254c0c6">

ryelle committed:
Remove focusout behavior

This was added to support the "click outside to close" behavior, but that can glitch when the click target is not a focusable element — clicking the space between buttons in the actions bar would close the dropdown. So the hasHover check was added to prevent that, but that brought in new glitches, where the focus could be reset to the first focusable element if hover was changed.

This PR removes all that in favor of using the backdrop element as a click target.

ryelle committed:
Remove unused class

ryelle committed:
Fix spacing on mobile modal header

ryelle committed:
Move keydown listener to parent container to catch focus on toggle button

ryelle committed:
Check for span before updating the text

ryelle committed:
Wrap options in a fieldset to associate collective title

This also adds a new label "title" which is used on the small screen modal too.
